pkg/vmprovider/providers/vsphere2/vmlifecycle/update_status.go

```diff
@@ -164,19 +160,36 @@ func getGuestNetworkStatus(
 		}
 	}
 
-	for i := range guestInfo.Net {
-		// Skip pseudo devices.
-		if guestInfo.Net[i].DeviceConfigId != -1 {
-			status.Interfaces = append(status.Interfaces, guestNicInfoToInterfaceStatus(i, &guestInfo.Net[i]))
+	if len(guestInfo.Net) > 0 {
+		var networkInterfaces []vmopv1.VirtualMachineNetworkInterfaceSpec
+		if vmCtx.VM.Spec.Network != nil {
+			networkInterfaces = vmCtx.VM.Spec.Network.Interfaces
 		}
-	}
 
-	// Hack: the exceedingly common case is just one nic - our boostrap effectively only works with one -
-	// so do the best effort here and apply the name in that common case. We have a long ways to go until
-	// we can always line up interfaces in the Spec with what both is observed in the VM hardware config
-	// and what the GuestNicInfo provides.
-	if len(networkInterfaces) == 1 && len(status.Interfaces) == 1 {
-		status.Interfaces[0].Name = networkInterfaces[0].Name
+		slices.SortFunc(guestInfo.Net, func(a, b types.GuestNicInfo) int {
+			return int(a.DeviceConfigId - b.DeviceConfigId)
+		})
+
+		ethCardIdx := 0
+		for i := range guestInfo.Net {
+			deviceKey := guestInfo.Net[i].DeviceConfigId
+
+			// Skip pseudo devices.
+			if deviceKey < 0 {
+				continue
+			}
+
+			var name string
+			if ethCardIdx < len(networkInterfaces) {
+				name = networkInterfaces[ethCardIdx].Name
+				ethCardIdx++
+			} else {
+				// Provide a default name that can otherwise hopefully help identify the interface.
+				name = fmt.Sprintf("nic-%d-%d", i, deviceKey)
```

Comment:
It is a little harder to use the same convention since we don't have an interface spec, so we don't have the interface name or the (optional) network name. We could default the interface name to ethX, but then we have to worry about collisions and whatnot. I'd prefer to see using the interface name for the name here in the status, and introduce the CR name as a object ref in v1a3+. 
I think I'll change this to make the device key num more clear and it have it be an invalid interface spec name, like `eth-deviceKey:12345` so we don't have to worry about a collision.